README.md
Chronus v0.1.0
==============

Go library and command line tool to help with time parsing, formatting, and accuracy


Features
--------

* [ ] Convert ISO 8160 datetime to
	* [x] UNIX Epoch timestamp
	* [x] SQL datetime string (ISO 8160 based)
	* [x] SQL datetime with timezone string (ISO 8160 based)
	* [x] RFC 3339 DateTime
	* [ ] Format (MM/DD/YYYY, YYYY-MM-DD HH:mm:ss TZ, etc.)
* [ ] Convert format (MM/DD/YYYY, YYYY-MM-DD HH:mm:ss TZ, etc.) date to
	* [ ] UNIX Epoch timestamp
	* [ ] ISO 8160 datetime string
	* [ ] RFC 3339 DateTime
	* [ ] SQL datetime string (ISO 8160 based)
	* [ ] SQL datetime with timezone string (ISO 8160 based)
* [ ] Convert UNIX Epoch timestamp to
	* [ ] ISO 8160 datetime string
	* [x] RFC 3339 DateTime
	* [ ] SQL datetime string (ISO 8160 based)
	* [ ] SQL datetime with timezone string (ISO 8160 based)
	* [ ] Format (MM/DD/YYYY, YYYY-MM-DD HH:mm:ss TZ, etc.)
* [ ] List file times
	* [ ] Access
	* [ ] Modification
	* [ ] Creation
	* [ ] etc.
* [ ] Option to read date from file modification time
* [ ] Parse time in the correct location
	* Automatically based on system settings?
	* When a country code is provided by the `CHRONUS_COUNTRY_CODE` environment variable or `-country-code` command line option
		* [ ] EU
		* [ ] UK
		* [x] US
		* [ ] etc.
* [ ] ____


Example Usage
-------------

### Parsing a Date Time With Timezone but Without Country Code Info

```bash
$ chronus "2021-03-08 16:06:34 MST"
               UNIX Timestamp: 1615219594
             Python Timestamp: 1615219594.000000
                 SQL DateTime: 2021-03-08 16:06:34
   UK Slash Date (DD/MM/YYYY): 08/03/2021
   US Slash Date (MM/DD/YYYY): 03/08/2021
            RFC 3339 DateTime: 2021-03-08T16:06:34Z
                     ISO 8601: 2021-03-08T16:06:34
           ISO 8601 Alternate: 2021-03-08 16:06:34 Z (MST)
ISO 8601 FileSafe 1 w/Seconds: 2021-03-08_160634Z
ISO 8601 FileSafe 2 w/Seconds: 2021-03-08_160634_MST

```

Note the default UTC timezone being referenced (the _Z_ instead of -0700 for MST) even though the numbers are correct. This is a Go issue.


### Parsing a Date Time With Timezone and Country Code Info

```bash
$ chronus -country-code US "2021-03-08 16:06:34 MST"
               UNIX Timestamp: 1615244794
             Python Timestamp: 1615244794.000000
                 SQL DateTime: 2021-03-08 16:06:34
   UK Slash Date (DD/MM/YYYY): 08/03/2021
   US Slash Date (MM/DD/YYYY): 03/08/2021
            RFC 3339 DateTime: 2021-03-08T16:06:34-07:00
                     ISO 8601: 2021-03-08T16:06:34
           ISO 8601 Alternate: 2021-03-08 16:06:34 -0700 (MST)
ISO 8601 FileSafe 1 w/Seconds: 2021-03-08_160634-0700
ISO 8601 FileSafe 2 w/Seconds: 2021-03-08_160634_MST

```

This does everything correctly because we supplied the right country code reference which is necessary as many countries often use the same abbreviated versions for timezones, i.e.; EST is use locally in most countries that have an Eastern Standard Time. Such as Australian Eastern Standard Time (AEST) is typically referred to as EST in Australia. So most timezone abbreviations can not be trusted.

Note that `CHRONUS_COUNTRY_CODE="US" chronus "2021-03-08 16:06:34 MST"` would have the same effect as the example above.

At this time "US" and "USA" are the only country codes supported. As there is interest, from myself or others, more country code support will be added. To have Chronus always default to US set the environment variable `CHRONUS_COUNTRY_CODE=US` or `CHRONUS_COUNTRY_CODE=USA`. Others countries will also be referenced via their standard Alpha-2 and Alpha-3 codes when their support is added.


Articles & Reference
--------------------

* [3.3. Date and Time Specification - RFC 5322][]
	* Obsoletes [3.3. Date and Time Specification - RFC 2822][]
		* Obsoletes [5. DATE AND TIME SPECIFICATION - RFC 822][]
* [List of TZ Database Time Zones - Wikipedia][]
* [Sources for time zone and daylight saving time data - IANA][]





[3.3. Date and Time Specification - RFC 5322]: https://datatracker.ietf.org/doc/html/rfc5322#page-14
[3.3. Date and Time Specification - RFC 2822]: https://datatracker.ietf.org/doc/html/rfc2822#page-14
[5. DATE AND TIME SPECIFICATION - RFC 822]: https://datatracker.ietf.org/doc/html/rfc822#section-5
[List of TZ Database Time Zones - Wikipedia]: https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
[Sources for time zone and daylight saving time data - IANA]: https://data.iana.org/time-zones/tz-link.html



CHANGELOG.md
ChangeLog:
==========

v1.0.0
------

Initial release



tzinfo/tzinfo.go
package tzinfo

import (
	"errors"
	"fmt"
	"log"
	"strconv"
	"strings"
	"time"
)

const (
	ErrorUnknownLocationStatus = "Unknown Location Status"
)

var (
	debug                       = false
	timeZoneDataByAbbreviation  map[string][]*TimeZoneLocation
	TimeZoneLocations           []*TimeZoneLocation
	unitedStatesAbbreviationMap map[string][]*TimeZoneLocation
)

type TimeZoneLocation struct {
	countryCodeAlpha2 string
	countryCodeAlpha3 string
	ianaName          string
	location          *time.Location
	nation            string
	offset            int
	status            string // timezone location status
	zone              string
}

// CountryCode returns the Alpha-2 (or Alpha-3 when necessary) country code if known
func (tzloc *TimeZoneLocation) CountryCode() string {
	// check if the Alpha-2 is set (not zero length) and not garbage (only two characters)
	if len(tzloc.countryCodeAlpha2) == 2 {
		return tzloc.countryCodeAlpha2
	}
	return tzloc.countryCodeAlpha3
}

// IANA returns the IANA time zone name
func (tzloc *TimeZoneLocation) IANA() string {
	return tzloc.ianaName
}

// Location returns a *time.Location for the time zone
func (tzloc *TimeZoneLocation) Location() *time.Location {
	if tzloc.location == nil {
		loc, err := time.LoadLocation(tzloc.ianaName)
		if err != nil {
			log.Printf("location error: %s\n", err.Error())
		} else {
			tzloc.location = loc
		}
	}

	return tzloc.location
}

// Nation returns the nation for the time zone if known
func (tzloc *TimeZoneLocation) Nation() string {
	return tzloc.nation
}

// Offset returns a formatted offset string. The default formatting is ±HHMM.
// For UTC/Zulu time is -0000. You can include modifiers to the zulu argument
// to adjust formatting. To add a colon for ±HH:MM add a colon to the
// zulu argument. To force a + instead of a - for UTC/Zulu time add one.
// To use use Z instead of a variation of -0000 just add it.
// Example: to specify ±HH:MM and use Z for UTC/Zulu time use the string ":Z".
// The order and capitalization of characters does not matter.
func (tzloc *TimeZoneLocation) Offset(zulu string) string {
	_, offset := time.Now().In(tzloc.location).Zone()
	return OffsetSecondsToString(offset, zulu)
}

// Status returns the current location status, one of alias, canonical, deprecated, or a zero length string if unset for the TZ Location
func (tzloc *TimeZoneLocation) Status() string {
	return tzloc.status
}

// StatusUpdate sets the current location status to one of alias, canonical, or deprecated
func (tzloc *TimeZoneLocation) StatusUpdate(s string) (err error) {
	status := strings.ToLower(s)
	switch status {
	case "", "alias", "canonical", "deprecated":
		tzloc.status = s
	default:
		err = errors.New(ErrorUnknownLocationStatus)
	}

	return err
}

// String returns a JSON encoded version of the data
func (tzloc *TimeZoneLocation) String() string {
	jsonStr := "{\n"

	jsonStr += `	"countryCodeAlpha2": "` + tzloc.countryCodeAlpha2 + `"` + ",\n"
	jsonStr += `	"countryCodeAlpha3": "` + tzloc.countryCodeAlpha3 + `"` + ",\n"
	jsonStr += `	"ianaName": "` + tzloc.ianaName + `"` + ",\n"
	jsonStr += `	"location": "` + tzloc.location.String() + `"` + ",\n"
	jsonStr += `	"nation": "` + tzloc.nation + `"` + ",\n"
	jsonStr += `	"offset": ` + strconv.Itoa(tzloc.offset) + ",\n"
	jsonStr += `	"status": "` + tzloc.status + `"` + ",\n"
	jsonStr += `	"zone": "` + tzloc.zone + `"` + ",\n"

	jsonStr = jsonStr[:len(jsonStr)-2] + "\n}\n"

	// byteSlice, err := json.Marshal(tzloc)
	// if err != nil {
	// 	return err.Error()
	// }
	// return string(byteSlice)
	return jsonStr
}

func (tzloc *TimeZoneLocation) Zone() (string, int) {
	return tzloc.zone, tzloc.offset
}

func init() {
	timeZoneDataByAbbreviation = make(map[string][]*TimeZoneLocation)
	unitedStatesAbbreviationMap = make(map[string][]*TimeZoneLocation)
	setTimeZoneLocation("AKDT", "-0900", "US/Alaska", "America")
	setTimeZoneLocation("AKST", "-1000", "US/Alaska", "America")
	setTimeZoneLocation("CDT", "-0500", "US/Central", "America")
	setTimeZoneLocation("CST", "-0600", "US/Central", "America")
	setTimeZoneLocation("EDT", "-0400", "US/Eastern", "America")
	setTimeZoneLocation("EST", "-0500", "US/Eastern", "America")
	setTimeZoneLocation("HDT", "-1000", "US/Aleutian", "America")
	setTimeZoneLocation("HST", "-1100", "US/Aleutian", "America")
	setTimeZoneLocation("HST", "-1100", "US/Hawaii", "America")
	setTimeZoneLocation("MDT", "-0600", "US/Mountain", "America")
	setTimeZoneLocation("MST", "-0700", "US/Mountain", "America")
	setTimeZoneLocation("PDT", "-0700", "US/Pacific", "America")
	setTimeZoneLocation("PST", "-0800", "US/Pacific", "America")
	setTimeZoneLocation("SST", "-1200", "US/Samoa", "America")
}

// DebugPrintf only prints output if tzinfo.debug is true
func DebugPrintf(f string, args ...interface{}) {
	if debug {
		log.Printf(f, args...)
	}
}

func GetCurrentTimeZoneLocation() *TimeZoneLocation {
	t := time.Now()
	zone, offset := t.Zone()

	// zulu := "Z"
	// offsetString := OffsetSecondsToString(offset, zulu)

	ianaLoc := "" // Somewhere/OverTheRainbow
	nation := ""  // Oz
	loc := time.FixedZone(zone, offset)

	tzloc := &TimeZoneLocation{
		zone:     zone,
		ianaName: ianaLoc,
		location: loc,
		nation:   nation,
		offset:   offset,
	}

	return tzloc
}

func GetNationByTZAbbreviation(abbr string) (string, error) {
	if data, ok := timeZoneDataByAbbreviation[abbr]; ok {
		return data[0].nation, nil
	}

	err := fmt.Errorf("zone %q not found in timezone data")
	return "", err
}

func GetTimeZoneLocationsByTZAbbreviation(abbr string) (loc []*TimeZoneLocation, err error) {
	if tzlocs, ok := timeZoneDataByAbbreviation[abbr]; ok {
		return tzlocs, err
	}

	err = fmt.Errorf("zone %q not found in timezone data")
	return nil, err
}

func GetUSTimeZoneLocationByTZAbbreviation(abbr string) (loc *TimeZoneLocation, err error) {
	if tzlocs, ok := unitedStatesAbbreviationMap[abbr]; ok {
		for i, tzloc := range tzlocs {
			DebugPrintf("chronus.GetUnitedStatesLocationByAbbreviation() | i: %d | tzloc: %q\n", i, tzloc)
			return tzloc, err
			break
		}
	}

	err = fmt.Errorf("zone %q not found in United States timezone data", abbr)
	return nil, err
}

func GetUSTimeZoneLocationsByTZAbbreviation(abbr string) (loc []*TimeZoneLocation, err error) {
	if tzlocs, ok := unitedStatesAbbreviationMap[abbr]; ok {
		return tzlocs, err
	}

	err = fmt.Errorf("zone %q not found in United States timezone data")
	return nil, err
}

func OffsetStringToSeconds(offsetStr string) (offset int) {
	// log.Printf("tzinfo.OffsetStringToSeconds() | offsetStr: %q\n", offsetStr)
	offsetStr = strings.TrimSpace(offsetStr)

	sign := '+'
	h10 := 'X'
	h01 := 'X'
	m10 := 'X'
	m01 := 'X'

	for _, r := range offsetStr {
		// log.Printf("tzinfo.OffsetStringToSeconds() | i: %d | r: %c\n", i, r)
		switch r {
		case '+', '-':
			// sign
			sign = r
		case ':':
			// colon ignored
		default:
			// digit
			if h10 == 'X' {
				h10 = r
			} else if h01 == 'X' {
				h01 = r
			} else if m10 == 'X' {
				m10 = r
			} else if m01 == 'X' {
				m01 = r
			} else {
				log.Fatalln("more than four digits in offset string")
			}
		}
	}

	hoursStr := fmt.Sprintf("%c%c", h10, h01)
	hoursInt, err := strconv.Atoi(hoursStr)
	if err != nil {
		// log.Printf("tzinfo.OffsetStringToSeconds() | err: %q\n", )
		log.Fatalln(err)
	}
	minutesStr := fmt.Sprintf("%c%c", m10, m01)
	minutesInt, err := strconv.Atoi(minutesStr)
	if err != nil {
		// log.Printf("tzinfo.OffsetStringToSeconds() | err: %q\n", err.Error())
		log.Fatalln(err)
	}

	if hoursInt > 0 {
		offset = hoursInt * 60 * 60
	}
	if minutesInt > 0 {
		offset += minutesInt * 60
	}
	if sign == '-' {
		offset = -offset
	}
	// log.Printf("tzinfo.OffsetStringToSeconds() | offset: %d\n", offset)

	offsetStr = OffsetSecondsToString(offset, "Z")
	// log.Printf("tzinfo.OffsetStringToSeconds() | offsetStr: %s\n\n", offsetStr)

	return offset
}

// OffsetSecondsToString returns a formatted offset string. The default formatting is ±HHMM. For UTC/Zulu time is -0000.
// You can include modifiers to the zulu argument to adjust formatting. To add a colon for ±HH:MM add one to the zulu argument.
// To force a + instead of a - for UTC/Zulu time add one. To use use Z instead of a variation of -0000 just add it.
// Example: to specify ±HH:MM and use Z for UTC/Zulu time use the string ":Z", or ":z". The order of characters does not matter.
func OffsetSecondsToString(offset int, zulu string) (offsetStr string) {
	colonStr := ""
	hours := 3600
	hourSeconds := 0
	minutes := 60
	minuteSeconds := 0
	positiveOffset := offset

	if offset == 0 {
		if len(zulu) > 0 {
			if strings.ContainsAny(zulu, "Zz") {
				return "Z"
			}

			result := "-00"
			if strings.Contains(zulu, "+") {
				result = "+00"
			}
			if strings.Contains(zulu, ":") {
				result += ":00"
			} else {
				result += "00"
			}
			return result
		}

		return "-0000"
	}

	if offset < 0 {
		positiveOffset = -offset
	}

	hourSeconds = positiveOffset / hours
	minuteSeconds = (positiveOffset - (hourSeconds * hours)) / minutes
	offsetStr = fmt.Sprintf("%02d%s%02d", hourSeconds, colonStr, minuteSeconds)

	DebugPrintf("tzinfo.OffsetSecondsToString() | offset: %d | positiveOffset: %d\n", offset, positiveOffset)
	DebugPrintf("tzinfo.OffsetSecondsToString() | hourSeconds: %02d\n", hourSeconds)
	DebugPrintf("tzinfo.OffsetSecondsToString() | minuteSeconds: %02d\n", minuteSeconds)

	if offset > 0 {
		offsetStr = "+" + offsetStr
	} else {
		offsetStr = "-" + offsetStr
	}

	return offsetStr
}

func setTimeZoneLocation(abbr, offset, ianaLoc, nation string) {
	var tzloc *TimeZoneLocation

	loc, err := time.LoadLocation(ianaLoc)
	if err != nil {
		log.Printf("location error: %s\n", err.Error())
	}

	nationUpper := strings.ToUpper(nation)
	switch nationUpper {
	case "AMERICA", "US", "USA":
		status := "canonical"
		switch loc.String() {
		case "America/Atka",
			"America/Fort_Wayne",
			"America/Indianapolis",
			"America/Knox_IN",
			"America/Louisville",
			"America/Shiprock",
			"Navajo",
			"US/Alaska",
			"US/Aleutian",
			"US/Arizona",
			"US/Central",
			"US/East-Indiana",
			"US/Eastern",
			"US/Hawaii",
			"US/Indiana-Starke",
			"US/Michigan",
			"US/Mountain",
			"US/Pacific":
			status = "deprecated"
		}
		tzloc = &TimeZoneLocation{
			countryCodeAlpha2: "US",
			countryCodeAlpha3: "USA",
			zone:              abbr,
			ianaName:          ianaLoc,
			location:          loc,
			nation:            nation,
			offset:            OffsetStringToSeconds(offset),
			status:            status,
		}
		unitedStatesAbbreviationMap[abbr] = append(unitedStatesAbbreviationMap[abbr], tzloc)
		timeZoneDataByAbbreviation[abbr] = append(unitedStatesAbbreviationMap[abbr], tzloc)
	default:
		tzloc = &TimeZoneLocation{
			zone:     abbr,
			ianaName: ianaLoc,
			location: loc,
			nation:   nation,
			offset:   OffsetStringToSeconds(offset),
		}
	}
	TimeZoneLocations = append(TimeZoneLocations, tzloc)
}


cmd/chronus/main.go
package main

import (
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"time"

	"github.com/runeimp/chronus"
	// "github.com/runeimp/chronus/tzinfo"
)

const (
	appName    = "Chronus"
	appVersion = "0.1.0"
	appLabel   = "Chronus v0.1.0"
)

const usage = `%s

Usage: %s [OPTIONS] [DATE_TIME]

OPTIONS:
`

var (
	countryCodePtr *string
	debugPtr       *bool
	helpPtr        *bool
	inputPtr       *bool
	iso8601Ptr     *bool
	labelPtr       *bool
	listPtr        *bool
	pythonPtr      *bool
	rfc3339Ptr     *bool
	sqlDateTimePtr *bool
	sqlPtr         *bool
	unixAllPtr     *bool
	unixFloatPtr   *bool
	versionPtr     *bool
	internetPtr    *bool
)

func main() {
	countryCodePtr = flag.String("country-code", "", "What country code should be used in calculations")
	debugPtr = flag.Bool("debug", false, "Display debugging info")
	helpPtr = flag.Bool("help", false, "Display this help info")
	inputPtr = flag.Bool("input", false, "Display the input referenced")
	iso8601Ptr = flag.Bool("iso8601", false, "Display time in ISO 8601 formats")
	labelPtr = flag.Bool("label", false, "Display label for single formats")
	listPtr = flag.Bool("list", false, "List all supported formats")
	pythonPtr = flag.Bool("python", false, "Display a Python timestamp")
	rfc3339Ptr = flag.Bool("rfc3339", false, "Display time in RFC 3339 formats")
	sqlPtr = flag.Bool("sql", false, "Display SQL Date Time Formats")
	sqlDateTimePtr = flag.Bool("sql-datetime", false, "Display a SQL DateTime")
	unixAllPtr = flag.Bool("unix-all", false, "Display time in UNIX formats")
	unixFloatPtr = flag.Bool("unix-float", false, "Display a UNIX floating point timestamp")
	versionPtr = flag.Bool("version", false, "Display version info")
	internetPtr = flag.Bool("web", false, "Display Internet Date Time Formats")

	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), usage, appLabel, filepath.Base(os.Args[0]))

		flag.VisitAll(func(f *flag.Flag) {
			optionName := fmt.Sprintf("-%s", f.Name)
			if f.DefValue == "" {
				fmt.Fprintf(flag.CommandLine.Output(), "  %-13s  %s (no default)\n", optionName, f.Usage)
			} else {
				fmt.Fprintf(flag.CommandLine.Output(), "  %-13s  %s (default: %v)\n", optionName, f.Usage, f.DefValue)
			}
		})
		fmt.Println()
	}

	flag.Parse()

	if *helpPtr {
		usageAndExit(0)
	}

	if *listPtr {
		chronus.ListFormats()
		os.Exit(0)
	}

	if *versionPtr {
		fmt.Println(appLabel)
		os.Exit(0)
	}

	if *debugPtr {
		chronus.Debug()
	}

	if len(flag.Args()) == 0 {
		// usageAndExit(0)
		t := time.Now()
		input := t.Format(time.RFC3339Nano)
		outputFormatBlocks(input)
		// printFormatInt64WithLabel("UNIX Timestamp", t.Unix())
		// printFormatFloat64WithLabel("Python Timestamp", chronus.PythonTimestamp(t))
		// printFormatStringWithLabel(t, "SQL DateTime", chronus.SQLDateTime)
		// printFormatStringWithLabel(t, "ISO 8601 Alternate", chronus.ISO8601alt)
		// printFormatStringWithLabel(t, "ISO 8601 FileSafe 1 w/Seconds", chronus.ISO8601file1Seconds)
		// printFormatStringWithLabel(t, "ISO 8601 FileSafe 2 w/Seconds", chronus.ISO8601file2Seconds)
		// fmt.Println()
		os.Exit(0)
	}

	// example := "Thu, 6 May 2021 12:46:12 -0700 (PDT)"
	// example = "Thu, 6 May 2021 12:46:12 -0700"
	// example = "Thu, 6 May 2021 12:46:12 PDT"

	for _, input := range flag.Args() {
		outputFormatBlocks(input)
	}
}

// stdError sends a formatted string to stderr
func stdError(f string, args ...interface{}) {
	fmt.Fprintf(os.Stderr, f, args...)
}

func outputFormatBlocks(input string) {
	var (
		err error
		t   time.Time
	)
	if len(*countryCodePtr) > 0 {
		chronus.CountryCode = *countryCodePtr
	}
	t, err = chronus.Parse(input)
	zName, zOffset := t.Zone()
	chronus.DebugPrintf("main.outputFormatBlocks() | t.Zone().name %q | .offset %d\n", zName, zOffset)
	if err != nil {
		fmt.Printf("Time Parse Error: %s\n", err.Error())
	}
	if *inputPtr {
		fmt.Printf("                        Input: %q\n", input)
		// fmt.Printf("                       Format: %q\n", format)
	}

	if *internetPtr {
		printFormatStringWithLabel(t, "RFC 3339 DateTime", chronus.RFC3339)
	}

	if *iso8601Ptr {
		// fmt.Printf("                     ISO 8601: %s\n", t.Format(chronus.ISO8601))
		// fmt.Printf("           ISO 8601 Alternate: %s\n", t.Format(chronus.ISO8601alt))
		// fmt.Printf("           ISO 8601 FileSafe1: %s\n", t.Format(chronus.ISO8601file1))
		// fmt.Printf("           ISO 8601 FileSafe2: %s\n", t.Format(chronus.ISO8601file2))
		// fmt.Printf("ISO 8601 FileSafe 1 w/Seconds: %s\n", t.Format(chronus.ISO8601file1Seconds))
		// fmt.Printf("ISO 8601 FileSafe 2 w/Seconds: %s\n", t.Format(chronus.ISO8601file2Seconds))
		// fmt.Printf("                     RFC 3339: %s\n", t.Format(chronus.RFC3339))

		printFormatStringWithLabel(t, "ISO 8601", chronus.ISO8601)
		printFormatStringWithLabel(t, "ISO 8601 Alternate", chronus.ISO8601alt)
		printFormatStringWithLabel(t, "ISO 8601 FileSafe 1", chronus.ISO8601file1)
		printFormatStringWithLabel(t, "ISO 8601 FileSafe 2", chronus.ISO8601file2)
		printFormatStringWithLabel(t, "ISO 8601 FileSafe 1 w/Seconds", chronus.ISO8601file1Seconds)
		printFormatStringWithLabel(t, "ISO 8601 FileSafe 2 w/Seconds", chronus.ISO8601file2Seconds)
		printFormatStringWithLabel(t, "RFC 3339", chronus.RFC3339)
		// printFormatStringWithLabel(t, "____", ____)
		// printFormatStringWithLabel(t, "____", ____)
		// printFormatStringWithLabel(t, "____", ____)
		fmt.Println()
	}

	if *pythonPtr {
		if *labelPtr {
			printFormatFloat64WithLabel("Python Timestamp", chronus.PythonTimestamp(t))
		} else {
			fmt.Printf("%f\n", chronus.PythonTimestamp(t))
		}
	}

	if *rfc3339Ptr {
		if *labelPtr {
			printFormatStringWithLabel(t, "RFC 3339 DateTime", chronus.RFC3339)
		} else {
			fmt.Printf("%s\n", t.Format(chronus.RFC3339))
		}
	}

	if *sqlPtr {
		fmt.Printf("                 SQL DateTime: %s\n", t.Format(chronus.SQLDateTime))
		fmt.Printf(" SQL DateTime Year to Seconds: %s\n", t.Format(chronus.SQLDateTimeYearToSecond))
		fmt.Printf("  SQL DateTime Year to Minute: %s\n", t.Format(chronus.SQLDateTimeYearToMinute))
		fmt.Printf("         SQL Date Year to Day: %s\n", t.Format(chronus.SQLDateYearToDay))
		fmt.Printf("       SQL Date Year to Month: %s\n", t.Format(chronus.SQLDateYearToMonth))
	}

	if *sqlDateTimePtr {
		if *labelPtr {
			fmt.Printf("                 SQL DateTime: %s\n", t.Format(chronus.SQLDateTime))
		} else {
			fmt.Println(t.Format(chronus.SQLDateTime))
		}
	}

	if *unixAllPtr {
		printFormatInt64WithLabel("UNIX Timestamp", chronus.UnixTimestamp(t))
		printFormatFloat64WithLabel("UNIX Floating Point Timestamp", chronus.UnixFloat(t))
		printFormatInt64WithLabel("UNIX Timestamp in Nanoseconds", chronus.UnixNano(t))
		printFormatStringWithLabel(t, "ANSI C DateTime", chronus.ANSIC)
		printFormatStringWithLabel(t, "Git DateTime", chronus.GitDateTime)
		printFormatStringWithLabel(t, "Ruby DateTime", chronus.RubyDateTime)
		printFormatStringWithLabel(t, "UNIX DateTime", chronus.UnixDateTime)
		printFormatStringWithLabel(t, "RFC850 DateTime", chronus.RFC850)
		printFormatStringWithLabel(t, "RFC1123 DateTime", chronus.RFC1123)
		printFormatStringWithLabel(t, "RFC1123Z DateTime", chronus.RFC1123Z)
		printFormatStringWithLabel(t, "RFC3339 DateTime", chronus.RFC3339)
		// printFormatStringWithLabel(t, "____", ____)
		// printFormatStringWithLabel(t, "____", ____)
		fmt.Println()
	}

	if *unixFloatPtr {
		fmt.Printf("UNIX Floating Point Timestamp: %f\n", chronus.UnixFloat(t))
	}

	switch {
	case *iso8601Ptr:
	case *pythonPtr:
	case *rfc3339Ptr:
	case *sqlPtr:
	case *sqlDateTimePtr:
	case *unixAllPtr:
	case *unixFloatPtr:
	case *internetPtr:
	default:
		printFormatInt64WithLabel("UNIX Timestamp", t.Unix())
		printFormatFloat64WithLabel("Python Timestamp", chronus.PythonTimestamp(t))
		printFormatStringWithLabel(t, "SQL DateTime", chronus.SQLDateTime)
		printFormatStringWithLabel(t, "UK Slash Date (DD/MM/YYYY)", chronus.UKSlashDate)
		printFormatStringWithLabel(t, "US Slash Date (MM/DD/YYYY)", chronus.USSlashDate)
		printFormatStringWithLabel(t, "RFC 3339 DateTime", chronus.RFC3339)
		printFormatStringWithLabel(t, "ISO 8601", chronus.ISO8601)
		printFormatStringWithLabel(t, "ISO 8601 Alternate", chronus.ISO8601alt)
		printFormatStringWithLabel(t, "ISO 8601 FileSafe 1 w/Seconds", chronus.ISO8601file1Seconds)
		printFormatStringWithLabel(t, "ISO 8601 FileSafe 2 w/Seconds", chronus.ISO8601file2Seconds)
		fmt.Println()
	}
}

func printFormatFloat64WithLabel(label string, f float64) {
	fmt.Printf("%29s: %f\n", label, f)
}

func printFormatInt64WithLabel(label string, d int64) {
	fmt.Printf("%29s: %d\n", label, d)
}

func printFormatStringWithLabel(t time.Time, label, format string) {
	fmt.Printf("%29s: %s\n", label, t.Format(format))
	// fmt.Printf("%29s: %s | format: %q\n", label, t.Format(format), format)
}

func usageAndExit(exitCode int) {
	flag.Usage()
	os.Exit(exitCode)
}
